doc/sendmail.1
.TH sendmail 1
.SH NAME
sendmail \- sendmail emulator interface for flatmailer
.SH SYNOPSIS
.B sendmail
[
.B flags
] [
.I recipients
] <
.I message
.SH DESCRIPTION
This program is a front end program for
.B flatmailer-inject
and
.BR flatmailer-smtpd .
It is used by programs that expect a
.I sendmail
interface for sending email.
After parsing the command-line arguments, this program executes either
.B flatmailer-inject
or
.B flatmailer-smtpd
depending on the presence of the
.I \-bs
option on the command-line.
See the documentation for
.B flatmailer-inject
for details on how messages are reformatted and queued.
.SH OPTIONS
.TP
.B \-B TYPE
.TP
.B \-C FILE
.TP
.B \-d DEBUG
.TP
.B \-h COUNT
.TP
.B \-i
.TP
.B \-L TAG
.TP
.B \-N DSN
.TP
.B \-n
.TP
.B \-O OPTION
.TP
.B \-o OPTION
.TP
.B \-p PROTOCOL
.TP
.B \-q TIME
.TP
.B \-R RETURN
.TP
.B \-U
.TP
.B \-V ENVID
.TP
.B \-v
.TP
.B \-X LOGFILE
Ignored for compatibility
.TP
.B \-bm
Read mail from standard input (default).
.TP
.B \-bp
List information about mail queue. This executes
.BR mailq .
.TP
.B \-bs
Use the SMTP protocol on standard input and standard output. This
executes
.BR flatmailer-smtpd .
.TP
.B \-F ADDRESS
Sets the full name of the sender.
.TP
.B \-f ADDRESS
Sets the envelope sender address.
.TP
.B \-r ADDRESS
An alternate and obsolete form of the \-f flag.
.TP
.B \-t
Read message for recipients and ignore the command-line arguments.
.SH SEE ALSO
flatmailer-inject(1),
flatmailer-smtpd(1)


ChangeLog
commit 6567de3d2f5936692fd2e272688386efa95f11fb
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Jan 22 11:02:45 2016 -0600

    Add a trivial autogen.sh script, to generate autoconf/automake files

 autogen.sh | 5 +++++
 1 file changed, 5 insertions(+)
 create mode 100644 autogen.sh

commit 471b1b42c931129f78a5054f323223549d0ab268
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Jan 22 10:51:09 2016 -0600

    Update dates in sources and README

 README                        | 4 ++--
 lib/address-old.cc            | 2 +-
 lib/address.cc                | 2 +-
 lib/base64.cc                 | 2 +-
 lib/canonicalize.cc           | 2 +-
 lib/cli++/clitest.cc          | 2 +-
 lib/cli++/main.cc             | 2 +-
 lib/cli++/messages.cc         | 2 +-
 lib/config_read.cc            | 2 +-
 lib/config_readint.cc         | 2 +-
 lib/config_readlist.cc        | 2 +-
 lib/config_syserr.cc          | 2 +-
 lib/errcodes.cc               | 2 +-
 lib/fdbuf/fdbuf.cc            | 2 +-
 lib/fdbuf/fdbuf.h             | 2 +-
 lib/fdbuf/fdbuf_copy.cc       | 2 +-
 lib/fdbuf/fdibuf.cc           | 2 +-
 lib/fdbuf/fdibuf.h            | 2 +-
 lib/fdbuf/fdibuf_mystring.cc  | 2 +-
 lib/fdbuf/fdibuf_netstring.cc | 2 +-
 lib/fdbuf/fdobuf.cc           | 2 +-
 lib/fdbuf/fdobuf.h            | 2 +-
 lib/fdbuf/fdobuf_chownmod.cc  | 2 +-
 lib/fdbuf/fdobuf_seek.cc      | 2 +-
 lib/fdbuf/fdobuf_signed.cc    | 2 +-
 lib/fdbuf/fdobuf_unsigned.cc  | 2 +-
 lib/fdbuf/tlsibuf.cc          | 2 +-
 lib/fdbuf/tlsibuf.h           | 2 +-
 lib/fdbuf/tlsobuf.cc          | 2 +-
 lib/fdbuf/tlsobuf.h           | 2 +-
 lib/hostname.cc               | 2 +-
 lib/makefield.cc              | 2 +-
 lib/mystring/count.cc         | 2 +-
 lib/mystring/iter.cc          | 2 +-
 lib/mystring/iter.h           | 2 +-
 lib/mystring/join.h           | 2 +-
 lib/mystring/mystring.h       | 2 +-
 lib/mystring/rep.h            | 2 +-
 lib/selfpipe.cc               | 2 +-
 lib/setenv.cc                 | 2 +-
 lib/tcpconnect.cc             | 2 +-
 protocols/protocol.cc         | 2 +-
 protocols/qmqp.cc             | 2 +-
 protocols/smtp.cc             | 2 +-
 protocols/tls_gnutls.cc       | 2 +-
 protocols/tls_none.cc         | 2 +-
 src/dsn.cc                    | 2 +-
 src/inject.cc                 | 2 +-
 src/mailq.cc                  | 2 +-
 src/queue.cc                  | 2 +-
 src/send.cc                   | 2 +-
 src/sendmail.cc               | 2 +-
 src/smtpd.cc                  | 2 +-
 53 files changed, 54 insertions(+), 54 deletions(-)

commit 1971c4506e336d60439ae4539b0742cea2a648fb
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Fri Jan 22 10:35:22 2016 -0600

    Remove done items from the TODO list

 TODO | 14 ++------------
 1 file changed, 2 insertions(+), 12 deletions(-)

commit 9819b7b97b42255e9f8f25d951d9ac91989c0066
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Thu Jan 21 16:08:03 2016 -0600

    Add a "check" make target and fix distcheck problems

 Makefile.am                    |  3 ---
 test/Makefile.am               | 18 +++++++++---------
 test/functions.in              | 22 +++++++++++++++++-----
 test/{runtests.in => runtests} |  9 +++------
 test/tests/dsn                 |  2 +-
 test/tests/mailq               |  4 ++--
 test/tests/send                |  2 +-
 test/tests/smtpd               |  1 -
 8 files changed, 33 insertions(+), 28 deletions(-)
 rename test/{runtests.in => runtests} (63%)
 mode change 100644 => 100755

commit 80260ff8cf0761bfa50598fe166ca965afaa3a85
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Thu Jan 21 16:45:56 2016 -0600

    Rework path handling to allow for testing anywhere
    
    This introduces a $NULLMAILER_TEST_PREFIX environment variable that may
    be used to override built-in paths to programs, the queue, etc.
    
    In particular:
    
    $NULLMAILER_TEST_PREFIX/bin is used for programs
    
    $NULLMAILER_TEST_PREFIX/conf is used for config files
    
    $NULLMAILER_TEST_PREFIX/queue is used for the queue directories
    
    $NULLMAILER_TEST_PREFIX/protocols is used for protocol programs
    
    $NULLMAILER_TEST_PREFIX/sbin is used for system programs (nullmailer-queue)
    
    These overrides are not used by programs such as mailq and
    nullmailer-queue when they are running setuid to avoid privilege
    escalation problems.

 lib/Makefile.am                           |  3 ++-
 lib/{config_readint.cc => config_path.cc} | 40 ++++++++++++++++++++++++-------
 lib/config_read.cc                        |  3 +--
 lib/config_readlist.cc                    |  3 +--
 lib/configio.h                            |  3 +++
 lib/defines.h                             | 11 +++++----
 lib/forkexec.cc                           | 21 +++++++---------
 lib/forkexec.h                            |  3 ++-
 lib/make_defines.sh                       | 13 ++++------
 src/mailq.cc                              | 14 ++++++-----
 src/queue.cc                              | 20 +++++++++++-----
 src/send.cc                               | 16 +++++++++----
 src/sendmail.cc                           |  5 ++--
 test/Makefile.am                          |  2 +-
 test/functions.in                         | 17 +++++++------
 test/runtests.in                          |  9 +------
 test/tests/inject/queue                   |  9 +++----
 test/tests/mailq                          |  4 ++--
 test/tests/queue/rewrite                  |  8 +++----
 test/tests/send                           |  4 ++--
 20 files changed, 120 insertions(+), 88 deletions(-)
 copy lib/{config_readint.cc => config_path.cc} (57%)

commit 0f4f095ff406eb8ece43732d61ada0ca4f501ec5
Author: Bruce Guenter <bruce@untroubled.org>
Date:   Thu Jan 21 16:30:27 2016 -0600

    dsn: Add controls to limit how much of the message to copy

 doc/nullmailer-dsn.1 | 14 ++++++++++++++
 src/dsn.cc           | 19 ++++++++++++++++---
 test/tests/dsn       | 22 +++++++++++++++++++++-
 3 files changed, 51 insertions(+), 4 deletions(-)


src/sendmail.cc
// flatmailer -- a simple relay-only MTA
// Copyright (C) 2016  Bruce Guenter <bruce@untroubled.org>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// You can contact me at <bruce@untroubled.org>.  There is also a mailing list
// available to discuss this package.  To subscribe, send an email to
// <flatmailer-subscribe@lists.untroubled.org>.

#include "config.h"
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "configio.h"
#include "fdbuf/fdbuf.h"
#include "defines.h"
#include "forkexec.h"
#include "setenv.h"
#include "cli++/cli++.h"

const char* cli_program = "sendmail";
const char* cli_help_prefix = "Flatmailer sendmail emulator\n";
const char* cli_help_suffix = 0;
const char* cli_args_usage = "[recipients] <message";
const int cli_args_min = 0;
const int cli_args_max = -1;
const bool cli_only_long = true;

enum mode { mode_normal, mode_mailq, mode_smtp };

static int o_dummyi;
static const char* o_dummys;
static const char* o_sender = 0;
static int o_mode = 0;
static char* o_from;
static int use_header = false;

cli_option cli_options[] = {
  { 'B', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  {  0, "bm",  cli_option::flag, mode_normal, &o_mode, sizeof(o_mode),
     "Read mail from standard input (default)", 0 },
  {  0, "bp",  cli_option::flag, mode_mailq, &o_mode,sizeof(o_mode),
     "List information about mail queue", 0 },
  {  0, "bs",  cli_option::flag, mode_smtp, &o_mode,sizeof(o_mode),
     "Handle SMTP commands on standard input", 0 },
  { 'C', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'd', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'F', 0,    cli_option::string, 0, &o_sender, sizeof(o_sender),
    "Set the full name of the sender", 0 },
  { 'f', 0,    cli_option::string, 0, &o_from, sizeof(o_from),
    "Set the envelope sender address", 0 },
  { 'h', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'i', 0,    cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi), "Ignored", 0 },
  { 'L', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'm', 0,    cli_option::flag,   0, &o_dummyi, sizeof(o_dummyi), "Ignored", 0 },
  { 'N', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'n', 0,    cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi), "Ignored", 0 },
  { 'O', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'o', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Set sendmail option, ignored", 0 },
  {  0, "em",  cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi),
     "Ignored", 0 },
  {  0, "ep",  cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi),
     "Ignored", 0 },
  {  0, "eq",  cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi),
     "Ignored", 0 },
  { 'p', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'q', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'R', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'r', 0,    cli_option::string, 0, &o_from, sizeof(o_from),
    "An alternate and obsolete form of the -f flag", 0 },
  { 't', 0,    cli_option::flag, 1, &use_header, sizeof(use_header),
    "Read message for recipients", 0 },
  { 'U', 0,    cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi), "Ignored", 0 },
  { 'V', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  { 'v', 0,    cli_option::flag, 0, &o_dummyi, sizeof(o_dummyi), "Ignored", 0 },
  { 'X', 0,    cli_option::string, 0, &o_dummys, sizeof(o_dummys), "Ignored", 0 },
  CLI_OPTION_END
};

bool setenvelope(char* str)
{
  char* at = strchr(str, '@');
  if(at) {
    *at = 0;
    setenv("FLATMAILER_HOST", at+1, 1);
  }
  setenv("FLATMAILER_USER", str, 1);
  return true;
}

int do_exec(const char* program, const char* xarg1, int argc, char* argv[])
{
  mystring path = program_path(program);

  const char* newargv[argc+3];
  newargv[0] = path.c_str();
  int j = 1;
  if (xarg1)
    newargv[j++] = xarg1;
  for(int i = 0; i < argc; i++)
    newargv[j++] = argv[i];
  newargv[j] = 0;

  execv(newargv[0], (char**)newargv);
  ferr << "sendmail: Could not exec " << program << '.' << endl;
  return 1;
}

int cli_main(int argc, char* argv[])
{
  if(o_sender)
    setenv("FLATMAILER_NAME", o_sender, 1);
  if(o_from)
    if(!setenvelope(o_from))
      return -1;
  switch (o_mode) {
  case mode_smtp:
    return do_exec("flatmailer-smtpd", 0, 0, 0);
  case mode_mailq:
    return do_exec("mailq", 0, 0, 0);
  default:
    return do_exec("flatmailer-inject", use_header ? "-b" : "-e", argc, argv);
  }
}
